cnxepub/tests/test_formatters.py

```diff
@@ -1376,3 +1376,263 @@ def test_annotate_autdetectparent(self, requests_get):
 </div>"""
         self.assertEqual(xmlpp(expected_content.encode('utf-8')).split(b'\n'),
                          xmlpp(etree.tostring(self.exercise)).split(b'\n'))
+
+class ExerciseTemplateTestCase(unittest.TestCase):
+    def format_html(self, html):
+        return xmlpp(html.encode('utf-8')).split(b'\n')
+
+    def test_renders_full_set(self):
+        # Test case for an exercise with a full set of features
+        exercise = {
+            "total_count": 1,
+            "items": [
+                {
+                "images": [],
+                "tags": [
+                    "type:practice",
+                    "all",
+                    "another-test-tag"
+                ],
+                "uuid": "a6581f0e-529b-4eaf-b174-563c7c49831b",
+                "group_uuid": "2c185140-d8dd-4116-83fa-f17783d0c2e3",
+                "number": 25499,
+                "version": 3,
+                "uid": "25499@3",
+                "published_at": "2021-08-04T18:43:46.007Z",
+                "nickname": "contmath98",
+                "url": "test-url", # Added in formatters
+                "required_context" : {  # Added in formatters
```

Comment:
These comments are easy to miss. Would it help to reduce the size of this JSON by removing unused fields (like `"images": []` and `"authors":`?

Also, it might help to inject the `url` and `required_context` fields in as part of the test.

Reply:
I agree it would be good to inject these as part of the test; I'm going to make sure `url` is covered the same way `required_context` is covered in earlier tests. The reason why I don't do that here is to be more modular; URL and required context come from that `exercise_callback_factory` function, which at this point is doing a lot of different things, and I wanted to make a separate function to interface with the template so that the template can be tested on its own. Let me know what you think.